### PR TITLE
Rnn scorer

### PR DESCRIPTION
**Pull request by SeBorgey**

### CODE REVIEW

**Comment by SeBorgey** [comment]
- решить, что делать с дампером
- странная ошибка с импортом адама

File: autointent/modules/scoring/_rnn.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+class RNNScorer(BaseScorer):
+    """Scorer based on RNN model for text classification."""
+
+    name = "rnn"
+    supports_multiclass = True
+    supports_multilabel = True
+
+    def __init__(
+        self,
+        rnn_config: RNNConfig | str | dict[str, Any] | None = None,
```
Comment (voorhs) [inline autointent/modules/scoring/_rnn.py]: Лучше делать не так.лусге чтобы все параметры были непосредственными параметрами конструктора. Без этого не будет удобно указывать серч спейс

File: autointent/modules/scoring/_rnn.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        self._model.eval()
+
+    def predict(self, utterances: list[str]) -> npt.NDArray[Any]:
+        """Predict probabilities for utterances."""
+        if not hasattr(self, "_model") or not hasattr(self, "_vocab"):
+            msg = "Model is not trained. Call fit() first."
+            raise RuntimeError(msg)
+
+        x = self._texts_to_sequences(utterances)
+        device = self.rnn_config.device or ("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (voorhs) [inline autointent/modules/scoring/_rnn.py]: Эта строчка повторяется, лучше в конструкторе сохранить в атрибуты

File: tests/modules/scoring/test_rnn.py
Comment (voorhs) [inline tests/modules/scoring/test_rnn.py:1]: Нужна проверка что после дамп лоад предикшены не поменялись

**Review by voorhs — CHANGES_REQUESTED** [review]

**Comment by voorhs** [comment]
Дампер уже реализован в пр про CNNScorer, но кажется он не работает до конца. Можете обсудить с Лерой и заколлабиться

File: autointent/modules/scoring/_rnn.py (hunk 1 above)
Comment (SeBorgey) [reply autointent/modules/scoring/_rnn.py]:
```

class RNNConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")
    model_name: str = Field("rnn", description="Name of the RNN model.")
    embed_dim: int = Field(128, description="Dimension of word embeddings.")
    hidden_dim: int = Field(512, description="Dimension of hidden states in RNN.")
    n_layers: int = Field(2, description="Number of RNN layers.")
    dropout: float = Field(0.1, description="Dropout rate.")
    device: str = Field(None, description="Torch notation for CPU or CUDA.")
    max_seq_length: int = Field(128, description="Maximum sequence length.")
    padding_idx: int = Field(0, description="Index used for padding.")
    pretrained_embs: Any = Field(None, description="Pretrained embedding weights if available.")
    batch_size: PositiveInt = Field(32, description="Batch size for model inference.")

    @classmethod
    def from_search_config(cls, values: dict[str, Any] | str | BaseModel | None) -> Self:
        if values is None:
            return cls()
        if isinstance(values, BaseModel):
            return values  # type: ignore[return-value]
        if isinstance(values, str):
            return cls(model_name=values)
        return cls(**values)

```
Я правильно понял, что я должен удалить обратно этот класс, и все параметры продублировать в самом классе везде, где нужно?

Comment (SeBorgey) [reply autointent/modules/scoring/_rnn.py]: тогда init станет очень большим и from_context тоже.
Одни и те же параметры будут перечисляться 3 раза подряд в коде. Сначала в аргументах init, потом в реализации init, затем в аргументах from_context.
ruff запрещает в аргументы функции пихать больше 10 параметров

Comment (voorhs) [reply autointent/modules/scoring/_rnn.py]: в целом верно, если что будем игнорить ошибку ruff, потому что серч спейс важнее

но я вот смотрю, некоторые параметры это не совсем гиперпараметры а действительно конфиг. я бы сделал так:
- отнести в конфиг: device, max_seq_length, padding_idx
- отнести в init: embed_dim, hidden_dim, n_layers, dropout
- убрать: model_name, pretrained_embs

**Review by voorhs — APPROVED** [review]
